File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-5.cs
public class Solution {
    public int LengthOfLongestSubstring(string s) {
        var left = 0;
        var charSet = new HashSet<char>();
        var maxCount = 0;

        for (var right = left; right < s.Length; right++) {
            while (charSet.Contains(s[right])) {
                charSet.Remove(s[left]);
                left++;
            }

            charSet.Add(s[right]);

            maxCount = Math.Max(maxCount, right - left + 1);
        }

        return maxCount;
    }
}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-5.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        seen = {}

        left = 0
        max_count = 0

        for right in range(len(s)):
            seen[s[right]] = 1 + seen.get(s[right], 0)

            length = right - left + 1
            most_freq = max(seen.values())

            replacements = length - most_freq

            while replacements > k:
                seen[s[left]] -= 1
                left += 1

                length = right - left + 1
                most_freq = max(seen.values())
    
                replacements = length - most_freq

            max_count = max(max_count, right - left + 1)

        return max_count


